api/modules/projects/controller.js

```diff
@@ -1,144 +1,158 @@
-    var record = req.pre.records.models[0];
+const ProjectsModel = require(`./model`);
 
-    return Publisher.publish(record)
-    .then(dateTracker.convertToISOStrings(true))
-    .then(function(publishedRecord) {
-      return req.generateResponse(publishedRecord).code(200);
-    });
-  })
-  .catch(errors.generateErrorResponse);
+const DateTracker = require(`../../../lib/utils`).DateTracker;
 
-  return reply(result);
-};
+class PublishedFilesCleanup {
+  constructor(project) {
+    this.project = project;
+  }
 
-controller.unpublishProject = function(req, reply) {
-  var result = Promise.resolve().then(function() {
-    var record = req.pre.records.models[0];
+  _fetchPublishedFiles(fileModel) {
+    return PublishedFilesModel.query({
+      where: {
+        file_id: fileModel.get(`id`)
+      }
+    })
+    .fetchAll();
+  }
 
-    if (!record.attributes.publish_url) {
-      throw Boom.notFound(null, {
-        debug: true,
-        error: 'This project was not published'
-      });
+  _deletePublishedFiles(publishedFilesModels) {
+    if (publishedFilesModels.length === 0) {
+      return Promise.resolve();
     }
 
-    return Publisher.unpublish(record)
-    .then(dateTracker.convertToISOStrings(true))
-    .then(function(unpublishedRecord) {
-      return req.generateResponse(unpublishedRecord).code(200);
+    return publishedFilesModels.mapThen(function(publishedFileModel) {
+      return publishedFileModel.destroy();
     });
-  })
-  .catch(errors.generateErrorResponse);
-
-  return reply(result);
-};
-
-controller.delete = function(req, reply) {
-  var self = this;
-  var project = req.pre.records.models[0];
-
-  function clearAllPublishedFiles() {
-    function fetchFiles() {
-      return Files.query({
-        where: {
-          project_id: project.get('id')
-        }
-      }).fetchAll();
-    }
+  }
 
-    function fetchPublishedFiles(file) {
-      return PublishedFiles.query({
-        where: {
-          file_id: file.get('id')
-        }
-      }).fetchAll();
+  cleanup() {
+    return FilesModel.query({
+      where: {
+        project_id: this.project.get(`id`)
+      }
+    })
+    .fetchAll()
+    .then(filesModels => {
+      return filesModels.mapThen(fileModel =>
+        this._fetchPublishedFiles(fileModel)
+        .then(this._deletePublishedFiles)
+      );
+    });
+  }
+}
+
+class ProjectsController extends BaseController {
+  constructor() {
+    super(ProjectsModel);
+  }
+
+  formatRequestData(request) {
+    const now = new Date();
+    const data = {
+      title: request.payload.title,
+      user_id: request.payload.user_id,
+      tags: request.payload.tags,
+      description: request.payload.description,
+      date_created: now,
+      date_updated: now,
+      readonly: request.payload.readonly,
+      client: request.payload.client
+    };
+
+    // If it is an update request
+    if (request.params.id) {
+      data.id = parseInt(request.params.id, 10);
+      delete data.date_created;
```

Comment:
unfortunately I can't do it out here since it looks like Bookshelf uses `hasOwnProperty` to determine if the key exists or not.